content/stdout-vs-stderr.md

```diff
@@ -0,0 +1,2077 @@
+    mem,
+    os::unix::io::{AsRawFd, FromRawFd},
+};
+
+fn read_metadata(file: &File) -> Result<()> {
+    let fd = file.as_raw_fd();
+    unsafe {
+        let _ = File::from_raw_fd(fd);
+    }
+    mem::forget(File::open("src/main.rs")?);
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    let mut file = File::open("Cargo.toml")?;
+
+    read_metadata(&file)?;
+
+    println!("Contents of Cargo.toml:\n{}", {
+        let mut s = String::new();
+        file.read_to_string(&mut s)?;
+        s
+    });
+    Ok(())
+}
+```
+
+</details>
+
+Moral of the story is, we need to put our code in a `unsafe` block like so:
+
+```rust
+use std::fs::File;
+
+// SAFETY: no other functions should call `from_raw_fd`, so there
+// is only one owner for the file descriptor.
+let raw_stdout = unsafe { File::from_raw_fd(1) };
+writeln!(raw_stdout, "test");
+```
+
+If you run it, you will see "test" on stdout. Yay! \o/
+
+However, as mentioned briefly in the previous section, there is still a big problem with this code. Going back to the documentation of `from_raw_fd`:
+
+> This function is typically used to **consume ownership** of the specified file descriptor. When used in this way, the returned object will take responsibility for **closing it** when the object goes out of scope.
+
+What this means is that `raw_stdout` variable takes ownership of the file descriptor and it will _close_ the stdout when it goes out of scope. In other words, when the created `File` is dropped, stdout is closed.
+
+<q>RIP.</q>
+
+We can confirm this behavior with this code:
+
+```rust
+use std::fs::File;
+use std::io::{Result, Write};
+use std::os::fd::FromRawFd;
+
+fn print1() -> Result<()> {
+    let mut raw_stdout = unsafe { File::from_raw_fd(1) };
+    writeln!(raw_stdout, "test1")
+}
+
+fn print2() -> Result<()> {
+    let mut raw_stdout = unsafe { File::from_raw_fd(1) };
+    writeln!(raw_stdout, "test2")
+}
+
+fn main() -> Result<()> {
+    print1()?;
+    print2()?;
+    Ok(())
+}
+```
+
+<center>
+<span class="glowy-code">
+
+[raw-stdout-broken.rs](https://github.com/orhun/rust-stdout-vs-stderr/blob/main/src/unbuffered/raw-stdout-broken.rs)
+
+</span>
+</center>
+
+What you expect to see is "test1" and "test2", however, stdout is closed after we leave the first function. When we try to open it again, it will panic because of the safety rule (the `fd` passed in must be an owned file descriptor + it must be **open**).
+
+```sh
+$ ./raw-stdout-broken.rs
+
+test1
+Error: Os { code: 9, kind: Uncategorized, message: "Bad file descriptor" }
+```
+
+<q>That's bad. What do we do?</q>
+
+In our case, we want the open file to _live through_ the entire program. Let's also assume that this is a TUI program and we have separate functions where passing the `raw_stdout` value around isn't possible.
+
+Well, there is still one quick dirty workaround: lazily initialize stdout and make it globally available via [`lazy_static`](https://docs.rs/lazy_static/latest/lazy_static/):
```

Comment:
I've been ditching lazy_static lately in favor of [once_cell](https://docs.rs/once_cell/latest/once_cell/), especially [`Lazy`](https://docs.rs/once_cell/latest/once_cell/#lazy-initialized-global-data).
I find it easier to use and this is what the compiler is recommending when you misuse a global const.

If I understand correctly it will be part of std at some point: https://www.reddit.com/r/rust/comments/126qaai/after_years_of_work_and_discussion_once_cell_has/

Not saying you should change it or anything but you might want to check it out.

Reply:
good point, added a note!